lnbits/tasks.py

```diff
@@ -181,3 +188,41 @@ async def invoice_callback_dispatcher(checking_id: str):
         for chan_name, send_chan in invoice_listeners.items():
             logger.trace(f"sse sending to chan: {chan_name}")
             await send_chan.put(payment)
+
+
+push_notification_key_file = os.path.join(
+    settings.lnbits_data_folder, "push_notification_vapid.{}"
+)
+
+
+def create_push_notification_keypair():
+    """
+    Generates VAPID key pair if not already exist
+    """
+    privkey = Vapid().from_file(push_notification_key_file.format("pem"))
+
+    if not os.path.exists(push_notification_key_file.format("pub")):
+        pubkey = privkey.public_key.public_bytes(
+            serialization.Encoding.X962, serialization.PublicFormat.UncompressedPoint
+        )
+        f = open(push_notification_key_file.format("pub"), "w")
+        f.write(b64urlencode(pubkey))
+        logger.info("✔️ Generated VAPID key pair for push notifications.")
+
+
+async def send_push_notification(subscription, title, body, url=""):
+    try:
+        logger.debug(f"sending push notification")
+        webpush(
+            json.loads(subscription.data),
+            json.dumps({"title": title, "body": body, "url": url}),
+            Vapid().from_file(push_notification_key_file.format("pem")),
+            {"aud": "", "sub": "mailto:devnull@example.com"},
+        )
+    except WebPushException as e:
+        """
+        This should clean up the database over time from stale endpoints
+        """
+        logger.error(f"failed sending push notification: {str(e)}")
+        await delete_push_notification_subscriptions(subscription.endpoint)
+        return
```

Comment:
not needed

Reply:
Yes, in most use cases this should not pile up to dimensions causing problems. Just thought having a simple cleanup routine is better than none at all, as they are not long living by nature and users will resubscribe sooner or later, producing new ones.

But such thing can be implemented when problems are actually experienced.

Reply:
Just found out about the WebPushException in case of unsubscribed or expired and could just catch the 410 Gone. There is no reason to hold on to them, they will be tried every time again and will always fail.
```
WebPushException: Push failed: 410 Gone
Response body:push subscription has unsubscribed or expired.
```